x/slashing/keeper_test.go

```diff
@@ -369,3 +368,107 @@ func TestHandleAlreadyJailed(t *testing.T) {
 	require.Equal(t, amtInt-1, validator.GetTokens().RoundInt64())
 
 }
+
+// Test a validator dipping in and out of the validator set
+// Ensure that missed blocks are tracked correctly and that
+// the start height of the signing info is reset correctly
+func TestValidatorDippingInAndOut(t *testing.T) {
```

Comment:
This test looks good - Might be worth splitting this up into two tests to keep distinct tests of different parts of the mechanism

Reply:
Yeah possibly - I think I'll leave it for this PR, but made an issue to restructure the testcases in this file more generally (this isn't the only large one...)

Reply:
https://github.com/cosmos/cosmos-sdk/issues/2499